### ISSUE TITLE
Performing surgery of a pre-trained model in CNTK

### ISSUE THREAD

**Issue by calledbymountains** [Microsoft/CNTK#2503]
Let us start by loading the [CNTK pretrained inception-v3 model][1].

There are following problems (or maybe I don't have a clear understanding) with the loaded model :

 1. The parameters are not properly ordered. In Inception-v3 model, there are many parallel blocks. However even if you load the [pre-trained AlexNet model][2] , you would not find any inherent ordering therein. 

> This raises the question that if someone provides a pre-trained model
> to me,he/she must also provide the model specification in terms of
> order of nodes and node names. Is that right ?

 2. If someone wants to modify a model, by let us say adding let us say a convolutional layer between 3rd and 4th convolutional layer, it is not very clear (again it might be my incomplete understanding) that how and where one should do that ? Should he/she squeeze between `'bias'` of 3rd layer and `'W'` of 4th layer ? 

In Caffe, one has a prototxt file which is helpful in clarifying these doubts. Unfortunately, I cannot find a clear specification/tutorial on these matters. 

Can someone kindly clarify ? I believe this will  help many people.

    In [7]: z = C.ops.functions.load_model('InceptionV3_ImageNet_CNTK.model')
    
    In [8]: z.parameters
    Out[8]: 
    (Parameter('W', [], [2048 x 1 x 1 x 1000]),
     Parameter('b', [], [1000]),
     Parameter('scale', [], [320]),
     Parameter('bias', [], [320]),
     Parameter('W', [], [320 x 2048 x 1 x 1]),
     Parameter('scale', [], [320]),
     Parameter('bias', [], [320]),
     Parameter('W', [], [320 x 1280 x 1 x 1]),
     Parameter('scale', [], [320]),
     Parameter('bias', [], [320]),
     Parameter('W', [], [320 x 192 x 3 x 3]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 288 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 288 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 256 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 192 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 80 x 3 x 3]),
     Parameter('scale', [], [80]),
     Parameter('bias', [], [80]),
     Parameter('W', [], [80 x 64 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 32 x 3 x 3]),
     Parameter('scale', [], [32]),
     Parameter('bias', [], [32]),
     Parameter('W', [], [32 x 32 x 3 x 3]),
     Parameter('scale', [], [32]),
     Parameter('bias', [], [32]),
     Parameter('W', [], [32 x 3 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 48 x 5 x 5]),
     Parameter('scale', [], [48]),
     Parameter('bias', [], [48]),
     Parameter('W', [], [48 x 192 x 1 x 1]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 96 x 3 x 3]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 64 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 192 x 1 x 1]),
     Parameter('scale', [], [32]),
     Parameter('bias', [], [32]),
     Parameter('W', [], [32 x 192 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 48 x 5 x 5]),
     Parameter('scale', [], [48]),
     Parameter('bias', [], [48]),
     Parameter('W', [], [48 x 256 x 1 x 1]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 96 x 3 x 3]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 64 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 256 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 256 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 48 x 5 x 5]),
     Parameter('scale', [], [48]),
     Parameter('bias', [], [48]),
     Parameter('W', [], [48 x 288 x 1 x 1]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 96 x 3 x 3]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 64 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 288 x 1 x 1]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 288 x 1 x 1]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 96 x 3 x 3]),
     Parameter('scale', [], [96]),
     Parameter('bias', [], [96]),
     Parameter('W', [], [96 x 64 x 3 x 3]),
     Parameter('scale', [], [64]),
     Parameter('bias', [], [64]),
     Parameter('W', [], [64 x 288 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 128 x 7 x 1]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 128 x 1 x 7]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 128 x 1 x 7]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 128 x 7 x 1]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 128 x 1 x 7]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 128 x 7 x 1]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 1 x 7]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 160 x 7 x 1]),
     Parameter('scale', [], [160]),
     Parameter('bias', [], [160]),
     Parameter('W', [], [160 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 7 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 1 x 7]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 1 x 7]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 7 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 1 x 7]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 7 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 3 x 3]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 7 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 192 x 1 x 7]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 768 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 1 x 3]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 1280 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 3 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 1 x 3]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 448 x 3 x 3]),
     Parameter('scale', [], [448]),
     Parameter('bias', [], [448]),
     Parameter('W', [], [448 x 1280 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 3 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 1280 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 1 x 3]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 2048 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 3 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 1 x 3]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 448 x 3 x 3]),
     Parameter('scale', [], [448]),
     Parameter('bias', [], [448]),
     Parameter('W', [], [448 x 2048 x 1 x 1]),
     Parameter('scale', [], [384]),
     Parameter('bias', [], [384]),
     Parameter('W', [], [384 x 384 x 3 x 1]),
     Parameter('scale', [], [192]),
     Parameter('bias', [], [192]),
     Parameter('W', [], [192 x 2048 x 1 x 1]),
     Parameter('W', [], [768 x 1 x 1 x 1000]),
     Parameter('b', [], [1000]),
     Parameter('scale', [], [768]),
     Parameter('bias', [], [768]),
     Parameter('W', [], [768 x 128 x 5 x 5]),
     Parameter('scale', [], [128]),
     Parameter('bias', [], [128]),
     Parameter('W', [], [128 x 768 x 1 x 1]))

  [1]: https://www.cntk.ai/Models/CNTK_Pretrained/InceptionV3_ImageNet_CNTK.model
  [2]: https://www.cntk.ai/Models/CNTK_Pretrained/AlexNet_ImageNet_CNTK.model

**Comment by cha-zhang** [comment]
For CNTK pre-trained models, we did provide the original script to train that. Some of these models were still trained with BrainScript, and the node names were not properly specified. We are in the process of updating these models with Python training. 

To add a layer between two convolution layers, in the case of Inception V3, there is also a batch normalization layer in between. So you want to add after 'scale' and before 'W'.

**Comment by calledbymountains** [comment]
@cha-zhang  Some of the original points were not clear and i am listing them below : 

- Assume a colleague trains a model and gives it to me. Now, I want to do surgery on it. As you can see, the parameters are not in the right order. They are jumbled. So, it is not directly possible to understand the network configuration. Would I need to visualize the model in TensorBoard to know the configuration ? Is there no way to understand the topology directly from the pre-trained model ? 
Please see the [original question](https://github.com/Microsoft/CNTK/issues/2503#issue-266199106) for details.

**Comment by cha-zhang** [comment]
You can plot the graph with: 
https://cntk.ai/pythondocs/cntk.logging.graph.html?highlight=plot#cntk.logging.graph.plot

**Comment by calledbymountains** [comment]
ok. So plotting the network graph is the only way.

**Comment by cha-zhang** [comment]
What would be an ideal solution for you? How would you prefer to understand the configuration of a network?

**Comment by calledbymountains** [comment]
No. I was just clarifying. It was not an objection.
